### ISSUE TITLE
Deploy API-X as a servlet filter?  

### ISSUE THREAD

**Issue by birkland** [fcrepo4-labs/fcrepo-api-x#103]
It has been discussed previously on API-X calls that it may be possible to expose API-X functionality as a servlet filter, e.g. intercepting and processing requests in a servlet filter pipeline in front of a repository webapp, as an alternative to deploying API-X as a separate piece of network infrastructure.   While there are availability and scalability downsides to doing so, it would present an even lower barrier to entry, allowing the creation of war files that bundle API-X and a repository webapp.  

This would likely involve:
* Verifying that this is indeed possible
* Add an OSGi framework to the war in order to deploy API-X components, or wire with Spring if not feasible
* Creating a servlet filter with hooks into the API-X components.
* Trying it out with the current Fedora impl (in a pattern similar to fcrepo-webapp-plus)

**Comment by acoburn** [comment]
I would note that while this would lower barriers to entry (an excellent goal!) and it would employ a traditional war-based deployment strategy, this also seems distinctly counter to most trends in application deployment methodologies. That is, rather than piling lots of applications into a single app server (tomcat, jetty, etc), most modern deployment strategies employ other, more scalable/distributed strategies, whether that involves an isolated Docker container or an OSGi container. As long as this move doesn't _prevent_ an OSGi (or containerless) deployment strategy, this is fine.

**Comment by ajs6f** [comment]
What @acoburn said, and also that what he says should happen _can_ happen, I believe. We may just be taling abot a bit (okay, a lot) of Maven cleverness.

**Comment by ajs6f** [comment]
👍 to the basic idea. This would be a _huge_ win for adoption, I believe.

**Comment by ajs6f** [comment]
> Using spring for wiring rather than OSGi (due to the absence of an OSGi framework)

Erm. grumble grumble. What about just loading an OSGi framework? They're really not that bulky or heavy. And Spring is... it's not the best thing ever.

**Comment by birkland** [comment]
@acoburn agreed.  JHU (for example) wouldn't have any specific interest in this deployment mode; the high-level stakeholder goals related to scalable/distributed deployments still stand!  I also see the docker-based demos as important educational tool which concretely demonstrates API-X as part of a microservice-style architecture in action.  That being said, it would be great if a single-war option serves as a stepping stone to help bootstrap adoption or evaluation.

@ajs6f Interesting... if it's tacking on an OSGi container can be done, that'd be great (preferable, even).  Edited description

**Comment by ajs6f** [comment]
Let's distinguish right now between an OSGi container (say, Karaf, or Eclipse Virgo) and an OSGi _framework_ (say, Apache Felix, or Eclipse Equinox). I'm talking about the latter. I am not talking about shoving Karaf into a webapp, I'm talking about Felix or Equinox, which are considerably smaller and lighter.
